src/managed/mod.rs

```diff
@@ -265,6 +263,7 @@ impl<T, E> Pool<T, E> {
 
         loop {
             obj.state = ObjectState::Receiving;
+            self.inner.available.fetch_sub(1, Ordering::Relaxed);
```

Comment:
When delaying this until the beginning of the loop the `available` counter doesn't get decreased for waiting tasks/workers. The original idea of `available` is that a negative number indicates that there are not enough connections available and there are `-available` waiting tasks/workers.

The only change needed is the other part of the patch you proposed. The count needs to be reduced by one if recycling fails. All other cases are already covered by the `Drop` implementation of the `Object`.